monte_carlo.py

```diff
@@ -79,26 +80,27 @@ def __init__(self, grid, gamma=0.9, num_episodes=10000, train_epsilon=0.9, test_
             self.policy.set_trained()
 
     @staticmethod
-    def get_legal_actions(state):
-        candidates = MonteCarlo.actions.values()
-        legal_actions = []
-        for c in candidates:
-            combined = (state[2] + c[0], state[3] + c[1])
-            if np.all(np.abs(combined) <= MonteCarlo.max_velocity) and np.any(combined):
-                legal_actions.append(c)
-
-        legal_actions = [list(MonteCarlo.actions.keys())[list(MonteCarlo.actions.values()).index(a)] for a in legal_actions]
-        return legal_actions
+    def get_valid_actions(state):
+        valid_actions = []
+        
+        if state not in MonteCarlo.actions_memoization:
+            for action, step in MonteCarlo.actions.items():
```

Comment:
```suggestion        
        if state not in MonteCarlo.actions_memoization:
            valid_actions = []
            for action, step in MonteCarlo.actions.items():
```

I think this scope makes more sense

Reply:
i though about this too; but was questioning about making the `action_memoization` a `defaultdict(list)`; but on second sight, i think this is a more self explanatory implementation